rand_core/src/lib.rs

```diff
@@ -212,6 +212,14 @@ pub trait RngCore {
 /// [`BlockRngCore`]: block::BlockRngCore
 pub trait CryptoRng: RngCore {}
 
+/// A marker trait used to indicate that an [`RngCore`] implementation is
+/// supposed to never return errors from the [`RngCore::try_fill_bytes`] method
+/// and never panic on unwrapping [`Error`] while calling other methods.
+///
+/// This trait is usually implemented by PRNGs, as opposed to OS, hardware,
+/// and periodically-reseeded RNGs, which may fail with IO errors.
+pub trait InfallibleRng: RngCore {}
```

Comment:
Commenting just for visibility of the significant addition.

It falls into the same trap as `std::iter::TrustedLen` in that it promises something about *another* trait.

I think I'm okay with this, but worth thinking about / getting more input on.

Reply:
I thought about different alternatives and it looks like it's the simplest option. Plus, we already have the `CryptoRng` trait.

One potential option is to make error an associated type (and `CryptoRng` may become associated `bool` constant). This way we could use [`Infallible`](https://doc.rust-lang.org/std/convert/enum.Infallible.html) by default. But without trait aliases it would be significantly less ergonomic to work with.

Reply:
Adding associated types + consts to `RngCore` changes the meaning on every existing usage as a generic bound.

For "crypto grade", I prefer the trait-inheritance model. After all, every `CryptoRng` is also a valid RNG. Also, can't have `CRYPTO_STRONG` default to `true` on all impls.

For infallibility, we'd at least need a bound on that error type: `Error: Into<Something>` where `Something` is either a boxed error or a summarized RNG error (`NotInitialized`, `EndOfSequence`, `IOError`, `MissingHardwareDevice`, ???).

Reply:
>every CryptoRng is also a valid RNG.

Yes, and? I don't see why the associated constant would be at odds with it. "Cryptographically strong" is a binary property of RNG, which can be modeled perfectly by an associated `bool` constant.

>Also, can't have CRYPTO_STRONG default to true on all impls.

Yes, it should've been `false` in my example.

>For infallibility, we'd at least need a bound on that error type

Sure. But I think it's an unimportant detail. Without stabilization of at least trait aliases, this approach is, arguably, not practical anyway.

Reply:
> Yes, and?

Only that this case does work well with the trait inheritance model; not every boolean property does.